src/v/cloud_storage/materialized_segments.cc

```diff
@@ -89,6 +113,22 @@ void materialized_segments::evict_segment(
     _cvar.signal();
 }
 
+ss::future<> materialized_segments::flush_evicted() {
+    if (_eviction_list.empty()) {
+        // Fast path, avoid waking up the eviction loop if there is no work.
+        co_return;
+    }
+
+    auto barrier = ss::make_lw_shared<eviction_barrier>();
+
+    // Write a barrier to the list and wait for the eviction consumer
+    // to reach it: this
+    _eviction_list.push_back(barrier);
+    _cvar.signal();
+
+    co_await barrier->promise.get_future();
```

Comment:
this is curious. i would have expected `co_return arrier->promise.get_future();`. it looks like the two versions would have the same semantics, but by returning the future immediately the caller gets to decide when to block the fiber.